runtime-common/current-run.ts

```diff
@@ -0,0 +1,769 @@
+    }
+
+    let localPath = this.#realmPaths.local(url);
+    let fileRef = await this.reader.readFileAsText(localPath);
+    if (!fileRef) {
+      return;
+    }
+    let { content, lastModified } = fileRef;
+    if (url.href.endsWith(".json")) {
+      let json = JSON.parse(content);
+      if (isCardDocument(json)) {
+        let instanceURL = new URL(url.href.replace(/\.json$/, ""));
+        if (opts?.delete && this.#instances.get(instanceURL)) {
+          this.#instances.remove(instanceURL);
+        } else {
+          json.data.id = instanceURL.href;
+          json.data.meta.lastModified = lastModified;
+          let module = await Loader.import<Record<string, any>>(
+            new URL(
+              json.data.meta.adoptsFrom.module,
+              new URL(localPath, this.realm.url)
+            ).href
+          );
+          let CardClass = module[json.data.meta.adoptsFrom.name] as typeof Card;
+          let card = CardClass.fromSerialized(json.data.attributes);
+          let searchData = await this.api.searchDoc(card);
+          this.#instances.set(instanceURL, {
+            resource: json.data,
+            searchData,
+            types: undefined,
+            refs: new Map(),
+          });
+        }
+      }
+    } else if (
+      hasExecutableExtension(url.href) &&
+      url.href !== `${baseRealm.url}card-api.gts` // the base card's module is not analyzable
+    ) {
+      if (opts?.delete) {
+        this.#modules.remove(url);
+        this.#modules.remove(trimExecutableExtension(url));
+      } else {
+        let mod = new ModuleSyntax(content);
+        this.#modules.set(url, mod);
+        this.#modules.set(trimExecutableExtension(url), mod);
+      }
+    }
+  }
+
+  private async semanticPhase(): Promise<void> {
+    for (let [url, mod] of this.modules) {
+      for (let possibleCard of mod.possibleCards) {
+        if (possibleCard.exportedAs) {
+          if (this.isIgnored(url)) {
+            continue;
+          }
+          await this.buildDefinition(
+            url,
+            mod,
+            {
+              type: "exportedCard",
+              module: url.href,
+              name: possibleCard.exportedAs,
+            },
+            possibleCard
+          );
+        }
+      }
+    }
+    for (let def of this.#definitions.values()) {
+      if (def.id.type !== "exportedCard") {
+        continue;
+      }
+      let { module } = def.id;
+      let refsMap = this.#exportedCardRefs.get(module);
+      if (!refsMap) {
+        refsMap = new Map();
+        this.#exportedCardRefs.set(module, refsMap);
+      }
+      let { type: remove, ...exportedCardRef } = def.id;
+      refsMap.set(internalKeyFor(def.id, undefined), exportedCardRef);
+    }
+
+    // once we have definitions we can fill in the instance types
+    for (let [url, entry] of [...this.instances]) {
+      entry.types = await this.getTypes(entry.resource.meta.adoptsFrom, url);
+    }
+
+    await this.buildRefsForInstances();
```

Comment:
this is very suboptimal and should definitely be refactored away as part of invalidation